test-network-function/accesscontrol/suite.go

```diff
@@ -130,16 +156,103 @@ func runTestOnPods(env *config.TestEnvironment, testCmd testcases.BaseTestCase,
+		crdPluralName := utils.ExecuteCommand(getCrPluralNameCommand, common.DefaultTimeout, common.GetContext(), func() {
+			tcClaimLogPrintf("CRD %s: Failed to get CR plural name.", crdName)
+		})
+		crNamespaces := getCrsNamespaces(crdName, crdPluralName)
+
+		ginkgo.By(fmt.Sprintf("CRD %s has %d CRs (plural name: %s).", crdName, len(crNamespaces), crdPluralName))
+		for crName, namespace := range crNamespaces {
+			ginkgo.By(fmt.Sprintf("Checking CR %s - Namespace %s", crName, namespace))
+			found := false
+			for _, configNamespace := range configNamespaces {
+				if namespace == configNamespace {
+					found = true
+				}
+			}
+
+			if !found {
+				tcClaimLogPrintf("CRD: %s (kind:%s) - CR %s has an invalid namespace (%s)", crdName, crdPluralName, crName, namespace)
+				if crNames, exists := invalidCrs[crdName]; exists {
+					invalidCrs[crdName] = append(crNames, crName)
+				} else {
+					invalidCrs[crdName] = []string{crName}
+				}
+			}
+		}
+	}
+	return invalidCrs
+}
+
 func testNamespace(env *config.TestEnvironment) {
-	ginkgo.When("test deployment namespace", func() {
+	ginkgo.When("test CNF namespaces", func() {
 		testID := identifiers.XformToGinkgoItIdentifier(identifiers.TestNamespaceBestPracticesIdentifier)
 		ginkgo.It(testID, func() {
-			for _, podUnderTest := range env.PodsUnderTest {
-				podName := podUnderTest.Name
-				podNamespace := podUnderTest.Namespace
-				ginkgo.By(fmt.Sprintf("Reading namespace of podnamespace= %s podname= %s, should not be 'default' or begin with openshift-", podNamespace, podName))
-				gomega.Expect(podNamespace).To(gomega.Not(gomega.Equal("default")))
-				gomega.Expect(podNamespace).To(gomega.Not(gomega.HavePrefix("openshift-")))
+			ginkgo.By(fmt.Sprintf("CNF resources' namespaces should not have any of the following prefixes: %v", invalidNamespacePrefixes))
+			var failedNamespaces []string
+			for _, namespace := range env.NameSpacesUnderTest {
+				ginkgo.By(fmt.Sprintf("Checking namespace %s", namespace))
+				for _, invalidPrefix := range invalidNamespacePrefixes {
+					if strings.HasPrefix(namespace, invalidPrefix) {
+						tcClaimLogPrintf("Namespace %s has invalid prefix %s", namespace, invalidPrefix)
+						failedNamespaces = append(failedNamespaces, namespace)
+					}
+				}
+			}
+			failedNamespacesNum := len(failedNamespaces)
+			if failedNamespacesNum > 0 {
+				ginkgo.Fail(fmt.Sprintf("Found %d namespaces with an invalid prefix.", failedNamespacesNum))
+			}
+
+			ginkgo.By(fmt.Sprintf("CNF pods' should belong to any of the configured namespaces: %v", env.NameSpacesUnderTest))
+			nonValidPodsNum := len(env.Config.NonValidPods)
+			if nonValidPodsNum > 0 {
+				for _, invalidPod := range env.Config.NonValidPods {
+					tcClaimLogPrintf("Pod %s has invalid namespace %s", invalidPod.Name, invalidPod.Namespace)
+				}
+
+				ginkgo.Fail(fmt.Sprintf("Found %d pods under test belonging to invalid namespaces.", nonValidPodsNum))
```

Comment:
I'm afraid this can happen in our partners lab. 
If partner is using multi namespace for CNF, then testing a single namespace only. There's a good chance the same labels are shared across namespaces and we will fail this test for no valid reason.

Reply:
I don't think labels would be shared across namespaces. It's possible though that they put all the labels in one config and only update the namespace list. Not sure if we really have to allow that. I don't think it's too much to ask them to update the label list as well.